### PR TITLE
fix(tasks): prevent position race conditions with atomic upserts and locking

### LINKED ISSUE #3089: *pq.Error: pq: duplicate key value violates unique constraint "uqe_task_positions_task_project_view"
Sentry Issue: [API-CLOUD-48](https://vikunja.sentry.io/issues/7588836146/?referrer=github_integration)

```
*pq.Error: pq: duplicate key value violates unique constraint "uqe_task_positions_task_project_view"
  File "/go/src/code.vikunja.io/api/pkg/routes/error_handler.go", line 134, in func1
  File "/go/src/code.vikunja.io/api/pkg/routes/error_handler.go", line 132, in reportToSentry
  File "/go/src/code.vikunja.io/api/pkg/routes/error_handler.go", line 108, in NewEcho.CreateHTTPErrorHandler.func4
...
(4 additional frame(s) were not displayed)
```

### PR DESCRIPTION
**Pull request by kolaente**
## Summary
Fixes race conditions in task position management that could violate the unique index on `(task_id, project_view_id)` when concurrent requests modify positions simultaneously. Introduces atomic upsert operations and explicit row locking to serialize concurrent position updates.

Fixes https://github.com/go-vikunja/vikunja/issues/3089

## Key Changes

- **New `upsertTaskPosition()` function**: Atomically inserts or updates a position row using database-native upsert syntax (`ON DUPLICATE KEY UPDATE` for MySQL, `ON CONFLICT DO UPDATE` for PostgreSQL). Replaces the separate exists-check + insert/update pattern that raced with concurrent requests.

- **New `insertTaskPositionsIgnoringExisting()` function**: Bulk-inserts position rows while gracefully skipping rows that already exist (using `ON DUPLICATE KEY UPDATE` or `ON CONFLICT DO NOTHING`). Batches inserts in groups of 100 to stay within database parameter limits. Used in paths where concurrent transactions may have created some rows in the meantime.

- **New `lockPositionsForViewUpdate()` function**: Takes an explicit row lock on the project view (via `SELECT ... FOR UPDATE`) to serialize concurrent transactions that rewrite positions for the same view. SQLite is excluded since it allows only a single writer at a time. This prevents two transactions from both deleting old position rows and then inserting overlapping new ones.

- **Updated `updateTaskPosition()`**: Now calls `lockPositionsForViewUpdate()` before upserting when position recalculation is needed, ensuring consistent lock ordering (view lock first, then position rows) to avoid deadlocks.

- **Updated `RecalculateTaskPositions()`**: Acquires the view lock at the start and deduplicates tasks in the result set (since the search can return a task more than once, but the unique index allows only one row per task/view pair).

- **Updated `recalculateTaskPositionsForRepair()`**: Acquires the view lock before repairing positions.

- **Updated `createPositionsForTasksInView()`**: Uses `insertTaskPositionsIgnoringExisting()` instead of a plain insert, allowing it to handle concurrent healing of the same tasks.

- **Updated `ensureTaskPositionsForSavedFilterView()`**: Removes the try-catch for unique constraint errors since `createPositionsForTasksInView()` now handles them internally.

- **Updated `UpdateTaskInSavedFilterViews` listener**: Uses `insertTaskPositionsIgnoringExisting()` to skip rows that may have been inserted concurrently.

## Tests
Added comprehensive test coverage for the new upsert and bulk-insert functions, including edge cases like updating existing rows and batching across multiple inserts.

https://claude.ai/code/session_01Rvn2MowBBaoxTVjAR9AV8K

### CODE REVIEW

**Comment by kolaente** [comment]
auggie review

**Comment by maximilize** [comment]
Confirming @augmentcode's line-221 concern — the residual race is real, and it reproduces the exact #3089 error from the recalc side.

The view lock only serializes writers that also take it, and `updateTaskPosition` skips it on the common path: `lockPositionsForViewUpdate` runs only when `tp.Position < MinPositionSpacing` (task_position.go:148-157), so a normal position set (`>= MinPositionSpacing`) upserts at :159 without holding it. Meanwhile `RecalculateTaskPositions` does `DELETE` + a plain `s.Insert(newPositions)` (task_position.go:305-312) — the one insert in this PR not switched to `insertTaskPositionsIgnoringExisting`. So an unlocked position set racing a recalc on the same view still collides on the unique index.

Deterministic repro on PG16 (table = the migration's unique index), U = unlocked upsert-insert of a task with no prior row, R = recalc:

```
R: BEGIN; SELECT id FROM project_views WHERE id=1 FOR UPDATE;
R: DELETE FROM task_positions WHERE project_view_id=1;      -- U's future row not visible yet
U: INSERT ... (100,1) ON CONFLICT DO UPDATE ...             -- no view lock; commits
R: INSERT ... (100,1) ...  (plain)
   ERROR: duplicate key value violates unique constraint "uqe_task_positions_task_project_view"
   DETAIL: Key (task_id, project_view_id)=(100, 1) already exists.
```

Minimal fix, consistent with the rest of the PR: give the `RecalculateTaskPositions` insert (:312) the same `ON CONFLICT (task_id, project_view_id) DO UPDATE SET position = excluded.position` upsert (its recomputed values should win after the delete), or take `lockPositionsForViewUpdate` on every `task_positions` writer rather than only the recalc-bound branch.

File: pkg/models/task_position.go [hunk 1]
```diff
@@ -154,6 +218,11 @@ func RecalculateTaskPositions(s *xorm.Session, view *ProjectView, a web.Auth) (e
 
 	log.Debugf("Recalculating task positions for view %d", view.ID)
 
+	err = lockPositionsForViewUpdate(s, view.ID)
```
[the remark this answers is not in the document]
Comment (kolaente) [reply pkg/models/task_position.go:248]: Good question — no, not all writers take the lock, and that's intentional: only the paths that delete-and-rewrite a whole view's positions (the two recalculation functions) need to be serialized against each other. The remaining writers (saved-filter healing, the task-created listener, the single-row upsert) only ever add or update individual rows and are all conflict-tolerant themselves (`ON CONFLICT DO NOTHING`/upsert), so they can't fail on the unique index.

The one residual race was the other direction: an unlocked writer slipping a row in between a recalculation's delete and its plain reinsert would have made the *recalculation* fail. Addressed in 6071af7 by making the recalculation's reinsert an authoritative upsert (`ON CONFLICT DO UPDATE`) — whatever an unlocked writer inserts concurrently just gets its position overwritten with the recalculated value, which is the correct outcome since the recalculation derives positions for every task in the view.

---
_Generated by [Claude Code](https://claude.ai/code)_
